src/Server.h

```diff
@@ -166,24 +165,53 @@ inline bool Server<IoType, SwitchActivityInfo, SwitchCalcGaps>::server_receive_t
 	bool do_update = true;
 	int ret = 0;
 	fds_data* l_fds_ifd = g_fds_array[ifd];
-	if (!l_fds_ifd)
+	if (!l_fds_ifd){
 		return (do_update);
-	ret = msg_recvfrom(ifd,
-			           l_fds_ifd->recv.cur_addr + l_fds_ifd->recv.cur_offset,
-			           l_fds_ifd->recv.cur_size,
-			           &recvfrom_addr);
-
-	if (ret == RET_SOCKET_SHUTDOWN) {
-		if (l_fds_ifd->sock_type == SOCK_STREAM) {
-			close_ifd( l_fds_ifd->next_fd,ifd,l_fds_ifd);
+	}
+#ifdef  USING_VMA_EXTRA_API
+	if (!g_vma_api || !g_pApp->m_const_params.is_vmapoll)
+#endif
+	{
+		ret = msg_recvfrom(ifd,
+				   l_fds_ifd->recv.cur_addr + l_fds_ifd->recv.cur_offset,
+				   l_fds_ifd->recv.cur_size,
+				   &recvfrom_addr);
+		if (ret == RET_SOCKET_SHUTDOWN) {
+			if (l_fds_ifd->sock_type == SOCK_STREAM) {
+				close_ifd( l_fds_ifd->next_fd,ifd,l_fds_ifd);
+			}
+			return (do_update);
 		}
-		return (do_update);
+		if (ret < 0) return (!do_update);
 	}
-	if (ret < 0) return (!do_update);
-
+#ifdef USING_VMA_EXTRA_API
 	int nbytes = ret;
-	while (nbytes) {
-
+	vma_buff_t* tmp_vma_poll_buff = g_vma_poll_buff;
+	while(tmp_vma_poll_buff || nbytes){
+		if (tmp_vma_poll_buff && !nbytes){
+			if (l_fds_ifd->recv.cur_offset) {
+				l_fds_ifd->recv.cur_addr = l_fds_ifd->recv.buf;
+				memcpy(l_fds_ifd->recv.cur_addr + l_fds_ifd->recv.cur_offset,(uint8_t*)tmp_vma_poll_buff->payload, tmp_vma_poll_buff->len);
+				recvfrom_addr = g_vma_comps->src;
+			}
+			else {
+				l_fds_ifd->recv.cur_addr = (uint8_t*)tmp_vma_poll_buff->payload;
+				recvfrom_addr = g_vma_comps->src;
+			}
+			ret = tmp_vma_poll_buff->len;
+			if (ret == RET_SOCKET_SHUTDOWN) {
+				if (l_fds_ifd->sock_type == SOCK_STREAM) {
+				  close_ifd( l_fds_ifd->next_fd,ifd,l_fds_ifd);
+				}
+				return (do_update);
+			}
+			if (ret < 0) return (!do_update);
+			nbytes = ret;
+		}
+#else
```

Comment:
see the same comment https://github.com/Mellanox/sockperf/pull/76/files?diff=unified#r76630541